wgpu-core/src/command/ray_tracing.rs

```diff
@@ -0,0 +1,1223 @@
+#[cfg(feature = "trace")]
+use crate::device::trace;
+#[cfg(feature = "trace")]
+use crate::ray_tracing::{
+    TlasInstance, TraceBlasBuildEntry, TraceBlasGeometries, TraceBlasTriangleGeometry,
+    TraceTlasInstance, TraceTlasPackage,
+};
+use crate::{
+    device::queue::TempResource,
+    global::Global,
+    id::CommandEncoderId,
+    init_tracker::MemoryInitKind,
+    lock::RwLockReadGuard,
+    ray_tracing::{
+        tlas_instance_into_bytes, BlasAction, BlasBuildEntry, BlasGeometries,
+        BuildAccelerationStructureError, TlasAction, TlasBuildEntry, TlasPackage,
+        ValidateBlasActionsError, ValidateTlasActionsError,
+    },
+    resource::{Blas, Tlas},
+    FastHashSet,
+};
+
+use wgt::{math::align_to, BufferAddress, BufferUsages, Features};
+
+use super::{BakedCommands, CommandBufferMutable};
+use crate::ray_tracing::BlasTriangleGeometry;
+use crate::resource::{AccelerationStructure, Buffer, Fallible, Labeled, StagingBuffer, Trackable};
+use crate::scratch::ScratchBuffer;
+use crate::snatch::SnatchGuard;
+use crate::storage::Storage;
+use crate::track::PendingTransition;
+use hal::BufferUses;
+use std::ops::Deref;
+use std::sync::Arc;
+use std::{cmp::max, num::NonZeroU64, ops::Range};
+
+type BufferStorage<'a> = Vec<(
+    Arc<Buffer>,
+    Option<PendingTransition<BufferUses>>,
+    Option<(Arc<Buffer>, Option<PendingTransition<BufferUses>>)>,
+    Option<(Arc<Buffer>, Option<PendingTransition<BufferUses>>)>,
+    BlasTriangleGeometry<'a>,
+    Option<Arc<Blas>>,
+)>;
+
+type BlasStorage<'a> = Vec<(
+    Arc<Blas>,
+    hal::AccelerationStructureEntries<'a, dyn hal::DynBuffer>,
+    u64,
+)>;
+
+// TODO: Get this from the device (e.g. VkPhysicalDeviceAccelerationStructurePropertiesKHR.minAccelerationStructureScratchOffsetAlignment) this is currently the largest possible some devices have 0, 64, 128 (lower limits) so this could create excess allocation (Note: dx12 has 256).
+const SCRATCH_BUFFER_ALIGNMENT: u32 = 256;
+
+impl Global {
+    pub fn command_encoder_build_acceleration_structures_unsafe_tlas<'a>(
+        &self,
+        command_encoder_id: CommandEncoderId,
+        blas_iter: impl Iterator<Item = BlasBuildEntry<'a>>,
+        tlas_iter: impl Iterator<Item = TlasBuildEntry>,
+    ) -> Result<(), BuildAccelerationStructureError> {
+        profiling::scope!("CommandEncoder::build_acceleration_structures_unsafe_tlas");
+
+        let hub = &self.hub;
+
+        let cmd_buf = hub
+            .command_buffers
+            .get(command_encoder_id.into_command_buffer_id());
+
+        let buffer_guard = hub.buffers.read();
+        let blas_guard = hub.blas_s.read();
+        let tlas_guard = hub.tlas_s.read();
```

Comment:
Oddly, we generally prefer not having long standing locks here - the hubs should have a `get` or similar, which will give you an Arc<T> directly, doing all the locking internally. This will also simplify the types you need to pass into various functions.

Reply:
I didn't realise those existed.